src/rest_vol.c

```
@@ -2732,14 +2477,258 @@ RV_copy_object_loc_info_callback(char *HTTP_response, void *callback_data_in, vo
+ * Function:    RV_copy_link_URI_by_index
+ *
+ * Purpose:     This callback is used to copy the URI of an link
+ *              in the server's response by index, to a provided buffer.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ * Programmer:  Matthew Larson
+ *              May, 2023
+ */
+herr_t
+RV_copy_link_URI_by_index(char *HTTP_response, void *callback_data_in, void *callback_data_out)
+{
+    yajl_val           parse_tree = NULL, key_obj = NULL, link_obj = NULL;
+    const char        *parsed_link_URI = NULL;
+    H5VL_loc_by_idx_t *idx_params      = (H5VL_loc_by_idx_t *)callback_data_in;
+    hsize_t            index           = idx_params->n;
+    char             **link_URI        = (char **)callback_data_out;
+    const char        *curr_key        = NULL;
+    herr_t             ret_value       = SUCCEED;
+
+    if (!HTTP_response)
+        FUNC_GOTO_ERROR(H5E_ARGS, H5E_BADVALUE, FAIL, "HTTP response buffer was NULL");
+
+    if (NULL == (parse_tree = yajl_tree_parse(HTTP_response, NULL, 0)))
+        FUNC_GOTO_ERROR(H5E_OBJECT, H5E_PARSEERROR, FAIL, "parsing JSON failed");
+
+    if (NULL == (key_obj = yajl_tree_get(parse_tree, links_keys, yajl_t_array)))
+        FUNC_GOTO_ERROR(H5E_OBJECT, H5E_PARSEERROR, FAIL, "failed to parse links");
+
+    if (key_obj->u.array.len == 0)
+        FUNC_GOTO_ERROR(H5E_OBJECT, H5E_PARSEERROR, FAIL, "parsed link array was empty");
+
+    if (index >= key_obj->u.array.len)
+        FUNC_GOTO_ERROR(H5E_OBJECT, H5E_PARSEERROR, FAIL, "requested link index was out of bounds");
+
+    switch (idx_params->order) {
+        case (H5_ITER_DEC):
+            if (NULL == (link_obj = key_obj->u.array.values[key_obj->u.object.len - 1 - index]))
+                FUNC_GOTO_ERROR(H5E_OBJECT, H5E_PARSEERROR, FAIL, "selected link was NULL");
+            break;
+
+        case (H5_ITER_NATIVE):
+        case (H5_ITER_INC):
+        case (H5_ITER_N):
```

Same comment as above where the `H5_ITER_N` and `H5_ITER_UNKNOWN` and `default` cases should return an error and only `H5_ITER_NATIVE` and `H5_ITER_INC` should have the below implementation.